### ISSUE TITLE
extrude_along_path with scaling crashes: AttributeError: 'numpy.ndarray' object has no attribute 'x'

### ISSUE THREAD

**Issue by tpchuckles** [SolidCode/SolidPython#151]
consider the creation of path, profile, and scaling lists sensibly and quickly via numpy:

`thetas=np.linspace(0,np.pi,N)
path=list(zip(3*np.sin(thetas),3*np.cos(thetas),thetas))
profile=list(zip(np.sin(thetas),np.cos(thetas)))
scalepts=list(np.linspace(1,.1,N))
`

vs building them element-by-element, without numpy:

`path=[] ; profile=[] ; scalepts=[]
for i in range(N):
	t=3.14/N*i
	path.append(Point3(3*math.sin(t),3*math.cos(t),t))
	profile.append(Point3(math.sin(t),math.cos(t),0))
	scalepts.append((N-i)/N)
`

the latter works, while the former yields error: "AttributeError: 'numpy.ndarray' object has no attribute 'x'", which is _especially_ confusing since i carefully converted my numpy arrays into lists! even when faced with the error, the user has no idea what's wrong (printing these lists shows nothing abnormal; python prints floats and numpy floats identically)

I would think the former example (minus the need to cast them into lists) would be preferred stylistically, especially if the latter case were to be broken into 3 for loops depending on the situation. the former will certainly be faster if these lists are large.

**Comment by etjones** [comment]
I’m away from a computer right now, but I think this is an issue with expecting the profile path to be a list of tuples, lists, or `euclid3.Point3`s, but not expecting the `np.ndarray `s that your code uses. It’s reasonable and increasingly common to use numpy to create geometry, so I’ll take a look at adding this capacity. For the moment, i suspect (but haven’t verified) that the line:

``` Python
profile = list(zip(map(math.sin, thetas), map(math.cos, thetas))) 
```
wouldn’t show the same error. 

There’s one function, `solid.utils.euclidify() ` that would would make your original (quite graceful, btw) code work if updated. I’ll see what would be required to make that fix.

**Comment by tpchuckles** [comment]
you're correct that the first code incorrectly presents tuples rather than euclid3.Point3 objects, but even when that is corrected (for i in range(N): profile[i]=Point3(profile[i][0],profile[i][1],profile[i][2])...etc, or via euclidify (thanks for that, that's awesome)), it still fails. 

they key to making it stop failing seemed to be in the building of scalepts list element-by-element. (I originally had posted a much different issue here, and since edited it, since this suggested the root of the issue was the numpy floats within the scalepts list). 

this code:

`1: path=euclidify(list(zip(3*np.sin(thetas),3*np.cos(thetas),thetas)))
2:	profile=euclidify(list(zip(np.sin(thetas),np.cos(thetas))))
3: 	scalepts=list(np.linspace(1,.1,N))

4: scalepts=[]
5: for i in range(N):
6:	scalepts.append((N-i)/N)
`

works with rows 4-6 present, fails with them commented out.

**Comment by etjones** [comment]
Thanks, this is great info. I'll see what I can figure out

**Comment by etjones** [comment]
Sorry for delay on this (I just had a baby! Little busy this week.) Looks like this is another numpy instance. Numpy has some different multiplication rules defined than standard python. Multiplying a standard python numeric by a `euclid3.Vector3 | euclid3.Point3` yields a `Vector3`.  Numpy's operator overrides supersede this, so multiplying a Numpy float64 by that same Vector3 yields a 3-tuple, which caused the problem you found. 

I'll push a fix for this behavior shortly, and then I'll see if I can add some more general behavior to `solid.utils.euclidify` that will play more nicely with Numpy.

**Comment by etjones** [comment]
Pushed and released as V1.0.1

**Comment by etjones** [comment]
Also worth noting: although you could call `euclidify()` on all the numpy lines like `path=list(zip(3*np.sin(thetas),3*np.cos(thetas),thetas))`, you don't have to. Once the issue with `scalepts` was fixed, this just works.
